# Working with text data

## Stage 1: Data preperation and samplig

We will discuss more on how data should be prepared for pre-tarining the LLM model. 

### Why we require embeddings?

- Deep neural network and also LLMs cannot process raw text data because they are mathematical models. 
- Therefore we need a way to represent word as a continuous valued vectors.
- The process of convertig texts into vectors called embeddings. 

## Let's build Tokenizer

### So, How we can make a tokenizer? (We will be building a tokenizer on small vocab and it is for learning purpose)

So what does tokenizer do? It takes the text string and converts into numerical Ids. 

Follwing are the steps to make our own tokenizer. 
1. Convert training text data into text tokens
2. Map those text tokens to numerical Ids which we can call VOCAB
3. Build Tokenizer class which takes the vocab and initialize encoder and decoder.
4. Encoder tranforms text string into tokenIds and then tokenIds can be converted into string again using decoder.



### Convert text string into text tokens

"Hello, I am here" -> ["Hello", ",", "I", "am", "here"]

We will use a texts store in training.txt file to get all training text data.

In [1]:
# we will be first load text file the-verdict.txt

with open("./assets/the-verdict.txt", "r") as file:
    text = file.read()

print(text)


I HAD always thought Jack Gisburn rather a cheap genius--though a good fellow enough--so it was no great surprise to me to hear that, in the height of his glory, he had dropped his painting, married a rich widow, and established himself in a villa on the Riviera. (Though I rather thought it would have been Rome or Florence.)

"The height of his glory"--that was what the women called it. I can hear Mrs. Gideon Thwing--his last Chicago sitter--deploring his unaccountable abdication. "Of course it's going to send the value of my picture 'way up; but I don't think of that, Mr. Rickham--the loss to Arrt is all I think of." The word, on Mrs. Thwing's lips, multiplied its _rs_ as though they were reflected in an endless vista of mirrors. And it was not only the Mrs. Thwings who mourned. Had not the exquisite Hermia Croft, at the last Grafton Gallery show, stopped me before Gisburn's "Moon-dancers" to say, with tears in her eyes: "We shall not look upon its like again"?

Well!--even through th

In [2]:
# we will convert that extracted text into text tokens 

tokens = text.split()

print(tokens)

# But we can see texts are only splitted by space. We need to split it by words and punctuations also includes 
# . , ! ? - _ @ "" ' ~ `` etc. and not whitespace

import re

tokens = re.findall(r"[^\W_]+|_|[^\w\s]", text, re.UNICODE)

print(tokens)


['I', 'HAD', 'always', 'thought', 'Jack', 'Gisburn', 'rather', 'a', 'cheap', 'genius--though', 'a', 'good', 'fellow', 'enough--so', 'it', 'was', 'no', 'great', 'surprise', 'to', 'me', 'to', 'hear', 'that,', 'in', 'the', 'height', 'of', 'his', 'glory,', 'he', 'had', 'dropped', 'his', 'painting,', 'married', 'a', 'rich', 'widow,', 'and', 'established', 'himself', 'in', 'a', 'villa', 'on', 'the', 'Riviera.', '(Though', 'I', 'rather', 'thought', 'it', 'would', 'have', 'been', 'Rome', 'or', 'Florence.)', '"The', 'height', 'of', 'his', 'glory"--that', 'was', 'what', 'the', 'women', 'called', 'it.', 'I', 'can', 'hear', 'Mrs.', 'Gideon', 'Thwing--his', 'last', 'Chicago', 'sitter--deploring', 'his', 'unaccountable', 'abdication.', '"Of', 'course', "it's", 'going', 'to', 'send', 'the', 'value', 'of', 'my', 'picture', "'way", 'up;', 'but', 'I', "don't", 'think', 'of', 'that,', 'Mr.', 'Rickham--the', 'loss', 'to', 'Arrt', 'is', 'all', 'I', 'think', 'of."', 'The', 'word,', 'on', 'Mrs.', "Thwing's",

In [3]:
# Now we need to turn this tokens into numerical tokenIds and map it. we will call it VOCAB 

text_tokens = sorted(set(tokens))

print(text_tokens)



['!', '"', "'", '(', ')', ',', '-', '.', ':', ';', '?', 'A', 'Ah', 'Among', 'And', 'Are', 'Arrt', 'As', 'At', 'Be', 'Begin', 'Burlington', 'But', 'By', 'Carlo', 'Chicago', 'Claude', 'Come', 'Croft', 'Destroyed', 'Devonshire', 'Don', 'Dubarry', 'Emperors', 'Florence', 'For', 'Gallery', 'Gideon', 'Gisburn', 'Gisburns', 'Grafton', 'Greek', 'Grindle', 'Grindles', 'HAD', 'Had', 'Hang', 'Has', 'He', 'Her', 'Hermia', 'His', 'How', 'I', 'If', 'In', 'It', 'Jack', 'Jove', 'Just', 'Lord', 'Made', 'Miss', 'Money', 'Monte', 'Moon', 'Mr', 'Mrs', 'My', 'Never', 'No', 'Now', 'Nutley', 'Of', 'Oh', 'On', 'Once', 'Only', 'Or', 'Perhaps', 'Poor', 'Professional', 'Renaissance', 'Rickham', 'Riviera', 'Rome', 'Russian', 'Sevres', 'She', 'Stroud', 'Strouds', 'Suddenly', 'That', 'The', 'Then', 'There', 'They', 'This', 'Those', 'Though', 'Thwing', 'Thwings', 'To', 'Usually', 'Venetian', 'Victor', 'Was', 'We', 'Well', 'What', 'When', 'Why', 'Yes', 'You', '_', 'a', 'abdication', 'able', 'about', 'above', 'abruptl

### Map text tokens to tokenIDs (It's also called vocab)

["Hi", "i"] -> [123, 456]

In [4]:
# we will create a dictionary to map tokens to tokenIds

vocab = {token: i for i, token in enumerate(text_tokens)}

print(vocab) # Dictionary of text to tokenIds is called vocab


{'!': 0, '"': 1, "'": 2, '(': 3, ')': 4, ',': 5, '-': 6, '.': 7, ':': 8, ';': 9, '?': 10, 'A': 11, 'Ah': 12, 'Among': 13, 'And': 14, 'Are': 15, 'Arrt': 16, 'As': 17, 'At': 18, 'Be': 19, 'Begin': 20, 'Burlington': 21, 'But': 22, 'By': 23, 'Carlo': 24, 'Chicago': 25, 'Claude': 26, 'Come': 27, 'Croft': 28, 'Destroyed': 29, 'Devonshire': 30, 'Don': 31, 'Dubarry': 32, 'Emperors': 33, 'Florence': 34, 'For': 35, 'Gallery': 36, 'Gideon': 37, 'Gisburn': 38, 'Gisburns': 39, 'Grafton': 40, 'Greek': 41, 'Grindle': 42, 'Grindles': 43, 'HAD': 44, 'Had': 45, 'Hang': 46, 'Has': 47, 'He': 48, 'Her': 49, 'Hermia': 50, 'His': 51, 'How': 52, 'I': 53, 'If': 54, 'In': 55, 'It': 56, 'Jack': 57, 'Jove': 58, 'Just': 59, 'Lord': 60, 'Made': 61, 'Miss': 62, 'Money': 63, 'Monte': 64, 'Moon': 65, 'Mr': 66, 'Mrs': 67, 'My': 68, 'Never': 69, 'No': 70, 'Now': 71, 'Nutley': 72, 'Of': 73, 'Oh': 74, 'On': 75, 'Once': 76, 'Only': 77, 'Or': 78, 'Perhaps': 79, 'Poor': 80, 'Professional': 81, 'Renaissance': 82, 'Rickham': 8

### Build Simple Tokenizer class which takes the vocab and initialize encoder and decoder.


In [5]:
class SimpleTokenizer:
    def __init__(self, vocab):
        self.str_to_id = vocab  
        self.id_to_str = {i: s for s, i in vocab.items()}

    def encode(self, text):
        # Convert text string into text tokens
        tokens = re.findall(r"[^\W_]+|_|[^\w\s]", text, re.UNICODE)
        return [self.str_to_id[token] for token in tokens]

    def decode(self, ids):
        return " ".join([self.id_to_str[id] for id in ids])


tokenizer = SimpleTokenizer(vocab)

encoded = tokenizer.encode("I HAD always thought Jack Gisburn rather a cheap genius")

print(encoded)

decoded = tokenizer.decode(encoded)

print(decoded) # we can see encoded tokens and if we pass encoded tokens to decoder it will give us back original text.


[53, 44, 149, 1012, 57, 38, 825, 115, 258, 492]
I HAD always thought Jack Gisburn rather a cheap genius


In [6]:
# But if we pass string which has some word which is not in vocab then it will throw an error.

# encoded = tokenizer.encode("Hello, I am here") # This will throw an error

# print(encoded)

That's why out vocab should contain maximum words so that encoder won't fail. We will see in next section how can we achieve our encoder not to fail.

### Adding special context tokens
In above example we saw if some words are not there in training data vocab then out tokenizer will throw error. How to counter that? one is make a dataset big to acquire all the words (That is not viable all the time) or there is another way to do it.

Add special context tokens like "<'endOfText'>" or "<'unk'>". These tokens we can add in vocab with token Id associated with it. if there is no text in training data while encoding then we will use token "<'unk'>". if there are two different string with different context then we will attach those strings with "<'endOfText'>" like e.g

Hi there I am villain.... You know there is charm in this weather ---> Hi there I am villain.... "<'endOfText'>" You know there is charm in this weather. Because two are different context sentences. Many tokenizer uses this types of special tokens to create encodings and decodings. 

we will extend previously created vocab to include "<'endOfText'>" and "<'unk'>"

In [7]:
#Rebuilding the vocab

preprocessed = sorted(set(tokens))

preprocessed.append("<endOfText>")
preprocessed.append("<unk>")

vocab = {token: i for i, token in enumerate(preprocessed)}

print(vocab) 


class SimpleTokenizer:
    def __init__(self, vocab):
        self.str_to_id = vocab  
        self.id_to_str = {i: s for s, i in vocab.items()}

    def encode(self, text):
        preprocessed = re.split(r'([,.:;?_!"()\']|--|\s)', text)
        preprocessed = [item.strip() for item in preprocessed if item.strip()]
        preprocessed = [item if item in self.str_to_id else "<unk>" for item in preprocessed]
        ids = [self.str_to_id[s] for s in preprocessed]
        return ids

    def decode(self, ids):
        text = " ".join([self.id_to_str[id] for id in ids])
        text = re.sub(r'\s+([,.:;?!"()\'])', r'\1', text)
        return text


tokenizer = SimpleTokenizer(vocab)

encoded = tokenizer.encode("I HAD always thought Jack Gisburn rather a cheap genius")

print(encoded)

decoded = tokenizer.decode(encoded)

print(decoded)

{'!': 0, '"': 1, "'": 2, '(': 3, ')': 4, ',': 5, '-': 6, '.': 7, ':': 8, ';': 9, '?': 10, 'A': 11, 'Ah': 12, 'Among': 13, 'And': 14, 'Are': 15, 'Arrt': 16, 'As': 17, 'At': 18, 'Be': 19, 'Begin': 20, 'Burlington': 21, 'But': 22, 'By': 23, 'Carlo': 24, 'Chicago': 25, 'Claude': 26, 'Come': 27, 'Croft': 28, 'Destroyed': 29, 'Devonshire': 30, 'Don': 31, 'Dubarry': 32, 'Emperors': 33, 'Florence': 34, 'For': 35, 'Gallery': 36, 'Gideon': 37, 'Gisburn': 38, 'Gisburns': 39, 'Grafton': 40, 'Greek': 41, 'Grindle': 42, 'Grindles': 43, 'HAD': 44, 'Had': 45, 'Hang': 46, 'Has': 47, 'He': 48, 'Her': 49, 'Hermia': 50, 'His': 51, 'How': 52, 'I': 53, 'If': 54, 'In': 55, 'It': 56, 'Jack': 57, 'Jove': 58, 'Just': 59, 'Lord': 60, 'Made': 61, 'Miss': 62, 'Money': 63, 'Monte': 64, 'Moon': 65, 'Mr': 66, 'Mrs': 67, 'My': 68, 'Never': 69, 'No': 70, 'Now': 71, 'Nutley': 72, 'Of': 73, 'Oh': 74, 'On': 75, 'Once': 76, 'Only': 77, 'Or': 78, 'Perhaps': 79, 'Poor': 80, 'Professional': 81, 'Renaissance': 82, 'Rickham': 8

In [8]:
encoded = tokenizer.encode("Hello, I am here") # This will throw an error

decoded = tokenizer.decode(encoded)
print(encoded)
print(decoded)

#Hello is not in vocab so it will be replaced with "<unk>"

text1 = "Hello, I am here"
text2=  "I am here"

text2encode = text1 + " <endOfText> " + text2

encoded = tokenizer.encode(text2encode)
decoded = tokenizer.decode(encoded)

print(text2encode)
print(encoded)
print(decoded)

# we can see that "<endOfText>" is added between text1 and text2.




[1140, 5, 53, 150, 542]
<unk>, I am here
Hello, I am here <endOfText> I am here
[1140, 5, 53, 150, 542, 1139, 53, 150, 542]
<unk>, I am here <endOfText> I am here


many people use different context tokens in there tokenizer fopr there LLMs. like <|BOS|> <|EOL|> or <|PAD|>. These are used in different contexts like some to understandf when does new sequence of tokens are getting generated. Some <|PAD|> like tokens for adding padding if text length is lesser, etc. 

But GPT 2 and GPT 3 uses only one token <|endoftext|>, it does not uses any other context token. Adding context token is more of providing context to LLMs about when does particular text is getting start or end. We will try to understand more in next section. But GPT 3.5 and GPT 4 uses some other token that we will see in next section

Moreover, GPT don't use <|unk|> type of context symbols for missing text in vocab. Rather it uses byte pair encoding. 

### Byte pair encoding
**What BPE is**: It is subword tokenization. Like rather than tokening the word "rare", BPE tokenize "ra" and "re" depending uplon the repeatness of particular subword. subwords are merged based on the frequency of it's in a training dataset. tiktoken has a BPE alogorithm and it is very complex to discuss it here in the scope of this learning. Rather than we will see how we can use it in encoding and decoding.

**Why it matters**: There will be some rare words while encoding some text string, and it is difficult to bring all types of words in vocab. Because while creating vocab it is necessary to make it suitable and light for tokenization.

**Workflow:** During training, count adjacent symbol pairs, merge the most frequent pair, and repeat until the vocabulary reaches the target size. At inference, the same merge rules encode text into token IDs.

You can see following figure to understand what is BPE

!['BPE Image'](./assets/bpe.png) 

In [9]:
import tiktoken

# cl100k_base is used by GPT-4 and GPT-3.5-turbo
encGPT3point5 = tiktoken.get_encoding("cl100k_base")
encGPT2 = tiktoken.get_encoding("gpt2")

print("GPT 3.5 special tokens:", encGPT3point5.special_tokens_set)
print("GPT 2 special tokens:", encGPT2.special_tokens_set)

text = "Hello, world! Byte pair encoding <|endoftext|> handles rare words and 你好."
token_ids = encGPT3point5.encode(text, allowed_special={"<|endoftext|>"})

print("Text:", repr(text))
print("Token IDs:", token_ids)
print("Round-trip:", encGPT3point5.decode(token_ids))
print("Number of tokens:", len(token_ids))


GPT 3.5 special tokens: {'<|endofprompt|>', '<|fim_suffix|>', '<|fim_prefix|>', '<|fim_middle|>', '<|endoftext|>'}
GPT 2 special tokens: {'<|endoftext|>'}
Text: 'Hello, world! Byte pair encoding <|endoftext|> handles rare words and 你好.'
Token IDs: [9906, 11, 1917, 0, 11146, 6857, 11418, 220, 100257, 13777, 9024, 4339, 323, 220, 57668, 53901, 13]
Round-trip: Hello, world! Byte pair encoding <|endoftext|> handles rare words and 你好.
Number of tokens: 17


As we can see GPT 3.5 uses cl100k_base vocab which has '<|fim_middle|>', '<|fim_suffix|>', '<|endofprompt|>', '<|endoftext|>', '<|fim_prefix|>' these special tokens but GPT 2 uses only one special token <|endoftext|> 

We can see token id of <|endoftext|> is 100257 which is too large and almost coming at the end of all tokens. In vocab spoecial tokens are stored at end of the dictionary. 

Note: You can also play with above code to check the token IDs of different diffferent text and also you can check how BPEs are being taken place for particular text. If there are more than one token Ids generated by encoder for single text then we can say it is going under BPE. 

### Data sampling with a sliding window

As we know tranformers are known to predict the next word or token if provided sequence of texts (tokens). so our next step is to generate a input-target pairs required for training an LLM

In next example we will see how to generate those input-target pairs for the `the-verdict` story

In [10]:
# Read texts from the-verdict.txt

import tiktoken

with open("./assets/the-verdict.txt", "r") as file:
    text = file.read()

# we will use tiktoken tokenizer to encode while story

enc = tiktoken.get_encoding("gpt2")

token_ids = enc.encode(text, allowed_special={"<|endoftext|>"})

# we will store two variables X and Y for input and target

X = []
Y = []

for i in range(len(token_ids) - 1):
    X.append(token_ids[i])
    Y.append(token_ids[i + 1])

print(f"X: {X}")
print(f"Y:    {Y}")

decoded_X = enc.decode(X)
decoded_Y = enc.decode(Y)

print(f"Decoded X: {decoded_X[:100]}")
print(f"Decoded Y:           {decoded_Y[:100]}")



X: [40, 367, 2885, 1464, 1807, 3619, 402, 271, 10899, 2138, 257, 7026, 15632, 438, 2016, 257, 922, 5891, 1576, 438, 568, 340, 373, 645, 1049, 5975, 284, 502, 284, 3285, 326, 11, 287, 262, 6001, 286, 465, 13476, 11, 339, 550, 5710, 465, 12036, 11, 6405, 257, 5527, 27075, 11, 290, 4920, 2241, 287, 257, 4489, 64, 319, 262, 34686, 41976, 13, 357, 10915, 314, 2138, 1807, 340, 561, 423, 587, 10598, 393, 28537, 2014, 198, 198, 1, 464, 6001, 286, 465, 13476, 1, 438, 5562, 373, 644, 262, 1466, 1444, 340, 13, 314, 460, 3285, 9074, 13, 46606, 536, 5469, 438, 14363, 938, 4842, 1650, 353, 438, 2934, 489, 3255, 465, 48422, 540, 450, 67, 3299, 13, 366, 5189, 1781, 340, 338, 1016, 284, 3758, 262, 1988, 286, 616, 4286, 705, 1014, 510, 26, 475, 314, 836, 470, 892, 286, 326, 11, 1770, 13, 8759, 2763, 438, 1169, 2994, 284, 943, 17034, 318, 477, 314, 892, 286, 526, 383, 1573, 11, 319, 9074, 13, 536, 5469, 338, 11914, 11, 33096, 663, 4808, 3808, 62, 355, 996, 484, 547, 12548, 287, 281, 13079, 410, 12523, 28

In [13]:
# Now we will create a training data of input string and target string
x = []
y = []

context_length = 10

for i in range(1, context_length):
    current_token_ids = token_ids[0:i]
    next_token_id = current_token_ids + [token_ids[i]]
    x.append(current_token_ids)
    y.append(next_token_id)

for i  in range(len(x)):
    print(f"{enc.decode(x[i])} ---------> {enc.decode(y[i])}")


I ---------> I H
I H ---------> I HAD
I HAD ---------> I HAD always
I HAD always ---------> I HAD always thought
I HAD always thought ---------> I HAD always thought Jack
I HAD always thought Jack ---------> I HAD always thought Jack G
I HAD always thought Jack G ---------> I HAD always thought Jack Gis
I HAD always thought Jack Gis ---------> I HAD always thought Jack Gisburn
I HAD always thought Jack Gisburn ---------> I HAD always thought Jack Gisburn rather


We can see how we have generated a input to target string. target strin g always have a next token. Which we can give to train LLM but before that now we have steps to remain to convert tose input-target mapping into embeddings. 

we will use pytorch tensors to create that embeddings. Embeddings helps to create context using vectors (tensors) for LLM.

If you need understanding on `pytorch` you can refer to `pytorch.ipynb` tutorial.